### DISCUSSION TITLE
Sorting by count of another collection based on relation

### DISCUSSION THREAD

**Discussion by jay-joo-code** [pocketbase/pocketbase#1770] [Q&A]
<img width="205" alt="image" src="https://user-images.githubusercontent.com/32003815/216773350-aacaeb0d-3a28-46ec-8ae7-8ab747e1cc8a.png">
<img width="671" alt="image" src="https://user-images.githubusercontent.com/32003815/216773355-aa0642db-757e-4511-b791-790af99648d5.png">

The question concerns two collections: topics, questions. 
Each question has one topic that it's associated to. 

Is there a way to fetch the top 10 topics that have the most number of questions related to it? 
It would be sorting topics by the number of questions pointing to it. 

https://github.com/pocketbase/pocketbase/discussions/1615 this seems like a similar question, but not exactly the same, since the collections in this post seem to be many-to-many.

**Comment by ganigeorgiev** [comment]
At the moment this is not possible without [extending PocketBase](https://pocketbase.io/docs/use-as-framework/), [registering a custom route](https://pocketbase.io/docs/router/#register-new-route) and [manually query the database](https://pocketbase.io/docs/database/).

Support for aggregations and custom queries are planned for the next v0.13/v0.14 release.
_Indirect relation filtering and sorting is also in the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1), but it could cause some performance overhead and I'll have to test it more thoroughly before introducing it._

**Reply by ganigeorgiev** [reply]
Here is an example, how this could look like if you decide to extend PocketBase (_I haven't tested the query, so you may have to adjust it_):

```go
package main

import (
    "log"
    "net/http"

    "github.com/labstack/echo/v5"
    "github.com/pocketbase/dbx"
    "github.com/pocketbase/pocketbase"
    "github.com/pocketbase/pocketbase/apis"
    "github.com/pocketbase/pocketbase/core"
    "github.com/pocketbase/pocketbase/daos"
    "github.com/pocketbase/pocketbase/models"
)

func FindTopTopics(dao *daos.Dao, limit int64) ([]*models.Record, error) {
    collection, err := dao.FindCollectionByNameOrId("topics")
    if err != nil {
        return nil, err
    }

    query := dao.RecordQuery(collection).
        AndSelect("count(q.id) as totalQuestions").
        InnerJoin("questions q", dbx.NewExp("q.topic = topics.id")).
        GroupBy("topics.id").
        OrderBy("totalQuestions DESC").
        Limit(limit)

    rows := make([]dbx.NullStringMap, 0, limit)

    if err := query.All(&rows); err != nil {
        return nil, err
    }

    records := models.NewRecordsFromNullStringMaps(collection, rows)

    return records, nil
}

func main() {
    app := pocketbase.New()

    app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
        e.Router.AddRoute(echo.Route{
            Method: http.MethodGet,
            Path:   "/top-topics",
            Handler: func(c echo.Context) error {
                records, err := FindTopTopics(app.Dao(), 10)
                if err != nil {
                    return apis.NewBadRequestError("Failed to fetch top topics", err)
                }

                // enable ?expand query param support
                apis.EnrichRecords(c, app.Dao(), records)

                return c.JSON(http.StatusOK, records)
            },
            Middlewares: []echo.MiddlewareFunc{
                apis.ActivityLogger(app),
                apis.RequireRecordAuth(),
            },
        })

        return nil
    })

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

**Reply by jay-joo-code** [reply]
thank you!
